### PR TITLE
Sort plios by name, created_at, updated_at, number_of_viewers

### LINKED ISSUE #173: Support for ordering plios (by multiple fields like name, date, number of viewers) and recieve a paginated response

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #173 

## Test Plan

- [x] Tested locally
- [x] Tested on staging
- [ ] Tested on production

### CODE REVIEW

File: plio/ordering.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    rest_framework/utils/filters.py
+
+    Two methods are overridden.
+    get_ordering(): Retrieves the ordering from the request query params,
+                    parses it into a list and checks it against a list of
+                    allowed ordering fields. If matched, returns the list
+
+    filter_queryset(): Gets the prepared ordering list from `get_ordering()`,
+                       prepares the queryset according to any specific custom
+                       ordering implementation, and returns the ordered by queryset
```
Comment (dalmia) [inline plio/ordering.py:19]:
```suggestion
                       ordering implementation, and returns the ordered queryset
```

File: plio/ordering.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    def get_ordering(self, request, queryset, view):
+        params = request.query_params.get(self.ordering_param)
+        if params:
+            fields = [param.strip() for param in params.split(",")]
+            # filter out the fields that are not allowed
+            ordering = [f for f in fields if f in self.allowed_ordering_fields]
+            if ordering:
+                return ordering
+
+        # No ordering was included, or all the ordering fields were invalid
```
Comment (dalmia) [inline plio/ordering.py:37]:
```suggestion
        # no ordering was included, or all the ordering fields were invalid
```

File: plio/ordering.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    # define the allowed custom ordering fields
+    # any ordering value other than this list will be rejected
+    allowed_ordering_fields = ["unique_viewers", "-unique_viewers"]
+
+    def get_ordering(self, request, queryset, view):
+        params = request.query_params.get(self.ordering_param)
+        if params:
+            fields = [param.strip() for param in params.split(",")]
+            # filter out the fields that are not allowed
+            ordering = [f for f in fields if f in self.allowed_ordering_fields]
```
Comment (dalmia) [inline plio/ordering.py:33]:
```suggestion
            ordering = [field for field in fields if field in self.allowed_ordering_fields]
```

File: plio/ordering.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        # No ordering was included, or all the ordering fields were invalid
+        return self.get_default_ordering(view)
+
+    def filter_queryset(self, request, queryset, view):
+        ordering = self.get_ordering(request, queryset, view)
+
+        if ordering:
+            # if the ordering fields contain "unique_viewers"
+            if any("unique_viewers" in orderBy for orderBy in ordering):
```
Comment (dalmia) [inline plio/ordering.py:45]: why not:

```
if "unique_viewers" in ordering
```

File: plio/ordering.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                # prepare a session queryset which has an annotated field "count_unique_users"
+                # that holds the count of unique users for every plio in the plio's queryset
+                session_queryset = Session.objects.filter(
```
Comment (dalmia) [inline plio/ordering.py:48]: variable name correction

File: plio/ordering.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                    plio__uuid=OuterRef("uuid")
+                ).values("plio__uuid")
+
+                session_queryset = session_queryset.annotate(
```
Comment (dalmia) [inline plio/ordering.py:52]: variable name correction

File: plio/ordering.py (hunk 5 above)
Comment (deepansh96) [inline plio/ordering.py:48]:
```suggestion
                plio_session_group = Session.objects.filter(
```

File: plio/ordering.py (hunk 6 above)
Comment (deepansh96) [inline plio/ordering.py:52]:
```suggestion
                count_unique_users_for_all_plios = plio_session_group.annotate(
```

File: plio/ordering.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                    count_unique_users=Count("user__id", distinct=True)
+                ).values("count_unique_users")
+
+                # annotate the plio's queryset with the count of unique users
+                queryset = queryset.annotate(
+                    unique_viewers=Coalesce(Subquery(session_queryset), 0)
```
Comment (deepansh96) [inline plio/ordering.py:58]:
```suggestion
                    unique_viewers=Coalesce(Subquery(count_unique_users_for_all_plios), 0)
```

**Pull request merged**